### PR TITLE
Deliverable E: Review for enhancement of the simulator on output files. GROUP D

### PR DESCRIPTION
**Pull request by manoj-goli**
Could you please review the final deliverable of our project? Below are few notes where we had made improvements to the existing simulator,

- In this deliverable, we have implemented an improvement to the existing simulator by following coding conventions and documentation. 
- Improvement: During the runtime of the program, the terminal will prompt the user to enter three inputs (i.e., start time, end time and component details) and these inputs will be used to print specific outputs i.e., the outputs within the timeframe and components specified by the user. User can also provide multiple components with a single space between them. 
- For example, if the user inputs start time as 00:00:00:000, end time as 00:20:00:000 and component as subnet1, then the output of the program will contain message details between the above timeframe and messages only through subnet1 (worked by Manoj and Revanth).

### CODE REVIEW

**Comment by gagandeep8030** [comment]
Please update the folder structure in the README.md file to include source and header files for PART E code.

**Comment by shubhamthakur463** [comment]
Kindly include instructions inside README.md to run and see the improvement in the simulator.

File: src/top_model/main.cpp [hunk 1]
```diff
@@ -225,6 +227,20 @@ std::shared_ptr<cadmium::dynamic::modeling::coupled<TIME>> TOP = std::make_share
     input = fopen("../data/output/abp_output.txt","r");
     output = fopen("../data/output/abp_processed_output.csv","w");
     filter(input,output);
-    
+    fclose(input);
+    fclose(output);
+    output_filter();
     return 0;
+}
+
```
Comment (gagandeep8030) [inline src/top_model/main.cpp]: Please provide comments for this newly written code (output_filter) for PART E.
The same comment applies to all the .cpp files.

**Comment by shubhamthakur463** [comment]
@manoj-goli @Revanth91 
Your project is working as expected but I am doubtful about using .csv file format. I am not sure if we can use this file format. Have you confirmed it with the professor?

**Comment by manoj-goli** [comment]
> @manoj-goli @Revanth91
> Your project is working as expected but I am doubtful about using .csv file format. I am not sure if we can use this file format. Have you confirmed it with the professor?

Hi,
Yes, @Revanth91 and I have checked with the professor earlier. Working with .CSV files is okay.
Thanks.

**Comment by shubhamthakur463** [comment]
> > @manoj-goli @Revanth91
> > Your project is working as expected but I am doubtful about using .csv file format. I am not sure if we can use this file format. Have you confirmed it with the professor?
> 
> Hi,
> Yes, @Revanth91 and I have checked with the professor earlier. Working with .CSV files is okay.
> Thanks.

Okay.

File: include/user_output_query.hpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+* provide multiple components with a single space between them. 
+*
+* For example, if the user inputs start time as 00:00:00:000, end time 
+* as 00:01:00:000 and component as subnet1, then the output of the program 
+* will contain message details between the above timeframe and messages 
+* only through subnet1.
+*
+* time			value	port				component
+* 00:00:20:000	1		packet_sent_out		sender1
+* 00:00:20:000	11		data_out			sender1
```
Comment (cruizm) [inline include/user_output_query.hpp]: The text is not aligned

File: src/top_model/main.cpp [hunk 3]
```diff
@@ -225,6 +227,20 @@ std::shared_ptr<cadmium::dynamic::modeling::coupled<TIME>> TOP = std::make_share
...
     output = fopen("../data/output/abp_processed_output.csv","w");
     filter(input,output);
-    
+    fclose(input);
+    fclose(output);
+    output_filter();
     return 0;
+}
+
+void output_filter(){
```
Comment (cruizm) [inline src/top_model/main.cpp]: Why this function is defined inside main.cpp?

File: src/user_output_query.cpp [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+* within the timeframe and components specified by the user. User can also 
+* provide multiple components with a single space between them. 
+*
+* For example, if the user inputs start time as 00:00:00:000, end time 
+* as 00:01:00:000 and component as subnet1, then the output of the program 
+* will contain message details between the above timeframe and messages 
+* only through subnet1.
+*
+* time			value	port				component
+* 00:00:20:000	1		packet_sent_out		sender1
```
Comment (cruizm) [inline src/user_output_query.cpp]: Same here. Please check all

**Comment by cruizm** [comment]
You may want to review the readme file to make sure that the instructions actually work. Are they for linux or for windows?

File: include/user_output_query.hpp (hunk 2 above)
Comment (manoj-goli) [reply include/user_output_query.hpp]: Corrected the mismatched alignments
Thank you.

File: src/user_output_query.cpp (hunk 4 above)
Comment (manoj-goli) [reply src/user_output_query.cpp]: Corrected.
Thanks!

**Comment by manoj-goli** [comment]
> You may want to review the readme file to make sure that the instructions actually work. Are they for linux or for windows?

Sure, we will check it,
Thanks

**Comment by revanth-sridhar** [comment]
> Please update the folder structure in the README.md file to include source and header files for PART E code.

This has been fixed, we request you to please check the files now.

**Comment by revanth-sridhar** [comment]
> Kindly include instructions inside README.md to run and see the improvement in the simulator.

Our improvement is to filter the outputs generated based on the inputs (such as start time, end time, component) entered by user during the runtime of the program.  We had provided details about it in the README file. Please check the attached image. If you feel, it requires further modifications, please let us know.
image

**Comment by revanth-sridhar** [comment]
> You may want to review the readme file to make sure that the instructions actually work. Are they for linux or for windows?

Many thanks for your feedback on the README file, we have updated the instructions. It should be good now.

File: src/top_model/main.cpp (hunk 3 above)
Comment (manoj-goli) [reply src/top_model/main.cpp]: Hi, 
Code has been updated in a better way, by removing the output_filter() function from all source files.
Thanks!

File: src/top_model/main.cpp (hunk 1 above)
Comment (revanth-sridhar) [reply src/top_model/main.cpp]: We have actually removed this function in the main.cpp files based on the reviews. However, we have provided new comments to the changes that we made to the main.cpp. May I request you to please check now?

**Review by revanth-sridhar — APPROVED** [review]
The improved suggested for the ABP simulator has been implemented by following the coding standards as part of the Deliverable E of the project.

**Review by shubhamthakur463 — APPROVED** [review]
Approved.

**Review by gagandeep8030 — APPROVED** [review]
Part E

**Pull request merged**